src/app/codegen-data-model/CodegenDataModel_Read.cpp

```diff
@@ -0,0 +1,350 @@
+/*
+ *    Copyright (c) 2024 Project CHIP Authors
+ *    All rights reserved.
+ *
+ *    Licensed under the Apache License, Version 2.0 (the "License");
+ *    you may not use this file except in compliance with the License.
+ *    You may obtain a copy of the License at
+ *
+ *        http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *    Unless required by applicable law or agreed to in writing, software
+ *    distributed under the License is distributed on an "AS IS" BASIS,
+ *    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *    See the License for the specific language governing permissions and
+ *    limitations under the License.
+ */
+#include <app/codegen-data-model/CodegenDataModel.h>
+
+#include <optional>
+#include <variant>
+
+#include <access/AccessControl.h>
+#include <access/Privilege.h>
+#include <access/RequestPath.h>
+#include <app-common/zap-generated/attribute-type.h>
+#include <app/AttributeAccessInterface.h>
+#include <app/AttributeAccessInterfaceRegistry.h>
+#include <app/AttributeValueEncoder.h>
+#include <app/GlobalAttributes.h>
+#include <app/RequiredPrivilege.h>
+#include <app/data-model/FabricScoped.h>
+#include <app/util/af-types.h>
+#include <app/util/attribute-metadata.h>
+#include <app/util/attribute-storage-detail.h>
+#include <app/util/attribute-storage-null-handling.h>
+#include <app/util/attribute-storage.h>
+#include <app/util/ember-global-attribute-access-interface.h>
+#include <app/util/ember-io-storage.h>
+#include <app/util/endpoint-config-api.h>
+#include <app/util/odd-sized-integers.h>
+#include <lib/core/CHIPError.h>
+#include <lib/support/CodeUtils.h>
+
+#include <zap-generated/endpoint_config.h>
+
+namespace chip {
+namespace app {
+namespace {
+using namespace chip::app::Compatibility::Internal;
+
+// Fetch the source for the given attribute path: either a cluster (for global ones) or attribute
+// path.
+//
+// if returning a CHIP_ERROR, it will NEVER be CHIP_NO_ERROR.
+std::variant<const EmberAfCluster *,           // global attribute, data from a cluster
+             const EmberAfAttributeMetadata *, // a specific attribute stored by ember
+             CHIP_ERROR                        // error, this will NEVER be CHIP_NO_ERROR
+             >
+FindAttributeMetadata(const ConcreteAttributePath & aPath)
+{
+    for (auto & attr : GlobalAttributesNotInMetadata)
+    {
+        if (attr == aPath.mAttributeId)
+        {
+            const EmberAfCluster * cluster = emberAfFindServerCluster(aPath.mEndpointId, aPath.mClusterId);
+            ReturnErrorCodeIf(cluster == nullptr, CHIP_IM_GLOBAL_STATUS(UnsupportedAttribute));
```

Comment:
This logic does not match the logic in the existing `ReadSingleClusterData`, and the result is that it's returning the wrong error code.

The error code for a supported global attribute will _never_ be UnsupportedAttribute, per spec.

I strongly suggest copying the "have a helper UnsupportedAttributeStatus function" bits from ember-compatibility-functions.cpp.

Reply:
https://github.com/project-chip/connectedhomeip/blob/master/src/app/util/ember-compatibility-functions.cpp#L305 seems to do this 

Overall I tried hard to keep compatibility with ember-compatibility-functions except that I found that its handling and usage of mExpanded to need updates.

Reply:
UnsupportedAttributeStatus never returns UnsupportedAttribute for a global attribute, right?

Reply:
Will re-read the logic to double-check.

Reply:
Ok, what threw me off I believe is that the UnsupportedAttributeStatus could return (for arbitrary path) a UnsupportedAttribute, however for globals we know that emberAfFindServerCluster is null so it will never return it.

Revised code, hopefully this is both readable and correct.